### PR TITLE
Add [GithubCheckRuns] service

### PR DESCRIPTION
**Pull request by mbtools**
## Description 

This service is for displaying the results of check runs in GitHub. 

## Data 

- Is there a public API? Yes
- Does the API require an API key? Uses typical GitHub auth approach
- Link to the API documentation: https://docs.github.com/en/rest/checks/runs#list-check-runs-for-a-git-reference

API path would be: `/{owner}/{repo}/commits/{ref}/check-runs`

For example: https://api.github.com/repos/badges/shields/commits/master/check-runs

## Motivation

Check Runs (and Check Suites) reflect the status of various GitHub Apps/Integrations/Services/etc. and are most often seen in the `Checks` tab for a PR or visible as icons next to the commit descriptions.

## Notes

This service returns the aggregated status of a ref. It currently does not allow to select a specific check run (job) as envisioned in #4364 but this can certainly be added later (as query parameter).

## Preview

image

image

### CODE REVIEW

**Comment by calebcartwright** [comment]
> It currently does not allow to select a specific check run (job) as envisioned in https://github.com/badges/shields/issues/4364 but this can certainly be added later (as query parameter).

Interesting, thanks for making this note! Since refs can be branches and branches can contain `/` that does mean we have to account for that in the route parameter (should be something like `:ref+`) though if we do keep the ref as a route parameter that prevents us from being able to incorporate any additional route params to follow

**Comment by mbtools** [comment]
It didn't occur to me that branches with `/` could be a problem but you are right. I use GH [check status](https://github.com/badges/shields/blob/master/services/github/github-checks-status.service.js#L13-L14) as a starting point which uses `:ref` but it seems like this is not correct either. Other places referring to branches use `:branch*` as the final part. 

Should it be `:ref+` or `:ref*`?

**Comment by calebcartwright** [comment]
> Should it be :ref+ or :ref*?

That determination is driven off the requirements of the upstream endpoint. Most of the time the upstream endpoint supports an optional parameter that can be a branch, but in a smaller set of cases the branch/ref is actually required.

I assumed that for check runs a ref would be required, is that correct?

**Comment by mbtools** [comment]
yes, ref is required

**Comment by calebcartwright** [comment]
> yes, ref is required

Then you'll want to use `+` to reflect that it's required (`*` would be optional/zero or more). I think the Bitbucket Pipelines badge is similar and may serve as a helpful reference

**Comment by calebcartwright** [comment]
I've done a quick pass through this and have opted to leave a few higher level comments/questions vs. inline comments.

Could you elaborate or provide some additional context on your thinking behind the mappings of quantities to messages? For example a count of zero being shown as 'no tests' is somewhat puzzling to me because checks are obviously more than tests.

As far as the code, seems to generally be in good order, though I'm curious if the various other helpers, e.g. `isBuildStatus` could be used more directly. If, not it'd be helpful if you could elaborate on the what and why of some of the custom mappings you're doing, e.g. a count of zero mapping to `no tests` considering that checks are much more than tests.

Additionally, I think more could be extracted within the `transform`, at least from the call flow perspective of the `handle` function; in cases where it makes sense to have a `transform` function (which indeed seems to be the case here) we'd typically expect whatever `transform` returns to be the thing that `handle` can pass along directly to whatever function is doing the rendering

**Comment by mbtools** [comment]
There are several status/conclusion values returned by the API that are not included in `build-status.js`. I didn't want to mess with the existing list there. 

The individual conclusions have to be aggregated in a way first and then mapped to a status anyway. I tried to map them to existing states that work with `renderBuildStatusBadge`. That's why it's `no tests`. `no check runs` would make more sense but requires changing `build-status.js`.

JS is really a hobby of mine so I'm sure there are more efficient ways to transform/map things. Happy to learn from someone who's more proficient than me.

**Comment by calebcartwright** [comment]
> There are several status/conclusion values returned by the API that are not included in build-status.js. I didn't want to mess with the existing list there.

Understood. There's always a balance between trying to determine whether to incorporate something from a service into the shared enumerations.

I still believe there's a bit more to think through here though relative to our desired behavior. For example, if the runs have an "in progress" status, does the conclusion field strictly always exist? Given the list of values defined in the schema for the conclusion, I'd be surprised if an actively running case has a conclusion field already set with one of those values. We should probably also look for opportunities to be consistent with our other similar types of badges (e.g. would a for the "in progress" state would message of `pending` be more consistent with other badges, like the check state, as opposed to using `processing` for the message)

On the implementation side, if we anticipate that we'll need a more service/endpoint-specific mapping beyond the general purpose build status (which I'm certainly open to, especially since it seems like we'll indeed need a two-fold mapping), then we should still refactor the code a bit to make it more easily testable. The `handle` function really should just be an orchestrator, and except in cases where the data extraction/transform is quite trivial (e.g. a simple destructure), we strongly prefer moving that logic outside the `handle` function so that it can be covered via more standard unit testing approaches.

**Review by mbtools — COMMENTED** [review]
Prettier

**Comment by mbtools** [comment]
I finally found some time for this...

Meanwhile GH has updated their API docs which made defining the joi schema a piece of cake. 

I combined the state mapping into a separate method and added a spec test for it as well. 

I ran `npm run prettier:check` and everything is fine. I can't figure out why it fails here. Can you have a look, please?

**Comment by mbtools** [comment]
One file had CR instead of LF... anyway, all checked have passed! Yay!

Ready for final review 😄

**Comment by mbtools** [comment]
bump

**Comment by jw3** [comment]
Looking forward to this one :rocket:

**Comment by mbtools** [comment]
the GH world is waiting for this one. would be great to get this merged :-)

**Comment by bourgeoisor** [comment]
This is exactly what I need! Is there anything blocking this PR?

I wanted to use `check-status` but since I have a few checks that don't _always_ run (neutral state) the badge is consistently set as "Pending.." even if no failures has occurred. This PR sounds like it fixes that issue!

**Comment by bourgeoisor** [comment]
What needs to happen for this PR to get some maintainers 👀 s on? It looks like it's been ready for a proper review since October (7 months ago). It would be nice to get this one in!

**Comment by mbtools** [comment]
Everything up-to-date and tested, again! 😃 

Devs are waiting (see above). It would be great to get a final review and merge, please @calebcartwright @chris48s

**Comment by calebcartwright** [comment]
@chris48s - just highlighting this one for us to potentially discuss or review. i know it keeps falling off my radar just because of how long it's been out here and queued up behind a number of blocked PRs, but afaik this one actually isn't blocked on anything other than a standard review

**Comment by mbtools** [comment]
Rules keep changing... Well, all checks pass (again) 😉

**Comment by calebcartwright** [comment]
Does the API that we're using work for a PR from a fork?

**Comment by mbtools** [comment]
According to GH docs: "The Checks API only looks for pushes in the repository where the check suite or check run were created. Pushes to a branch in a forked repository are not detected"

**Comment by calebcartwright** [comment]
Right, I'd just read that but wasn't sure if there was any subtly or nuance to it (something not uncommon in GitHub docs). That's not a limitation I was aware of when I opened #4364, and I confess it has me second guessing the benefits of this badge as currently proposed/structured.

What effect do people feel that has on the potential use cases for this badge? I.e. I think that means this badge will be mostly confined to post-merge scenarios (though I recognize it could ostensibly work for in-repo PRs).

Finally, and assuming there is some to-be-realized value scenario with this badge, does it seem best suited as an aggregate-based binary type badge (e.g. they all pass and green or one or more passed and red), or would it be better as a summary style badge, similar to what we offer with test result badges?

**Comment by mbtools** [comment]
We have been using this API for a while and it works well for in repo PRs. If you want the check-runs of a fork, I assume you just use the URL of the fork. 

The badge is intended to work the same way as the green checkmark or red X you see behind the description of pull requests. This aggregates the various checks. There is, however, an intermediate state when the checks are running. So not binary.

**Comment by daxpedda** [comment]
> The badge is intended to work the same way as the green checkmark or red X you see behind the description of pull requests. This aggregates the various checks.

This is also what I've been using it for so far.

**Comment by calebcartwright** [comment]
> We have been using this API for a while and it works well for in repo PRs. If you want the check-runs of a fork, I assume you just use the URL of the fork.

For Shields to offer a native badge/dedicated route it needs to be generally applicable to a large portion of our audience, and not just well suited to narrow use cases/scenarios. If something is going to create more questions/problems for us, then we often find that's better suited as a [Custom Endpoint badge](https://shields.io/badges/endpoint-badge).

I appreciate that it works well for you, and your specific context where you've got mostly/perhaps exclusive in-repo PRs (and that's why I mentioned this in my prior comment), but I hope you can also recognize that's not ubiquitous across public open source projects that shields.io supports.

I don't have a strongly held opinion one way or another, and this is why I am asking and why I'd like broader input on whether this type of badge fits a broader need/desire.

>  The badge is intended to work the same way as the green checkmark or red X you see behind the description of pull requests. This aggregates the various checks. There is, however, an intermediate state when the checks are running. So not binary.

I am familiar with the Shields code and our coloring paradigms, and I don't want to get caught up in pedantry over a simple example which fundamentally skips the question that I asked.

I am posing an open question about the value/interest in an aggregate, single status badge, especially since that's not what was proposed/discussed in the associated issue of #4364. I am asking whether this is something our users would want, and/or whether there would be interest in an aggregate, summary badge (e.g. , )

Note: these are not mutually exclusive options, both could be on the table if there's compelling levels of interest in either/both

**Comment by daxpedda** [comment]
> I am posing an open question about the value/interest in an aggregate, single status badge, especially since that's not what was proposed/discussed in the associated issue of #4364. I am asking whether this is something our users would want, and/or whether there would be interest in an aggregate, summary badge (e.g. , )

Personally I think badges that show number of checks or show passed/failed checks is pretty niche.

In contrast, having a badge that simply shows if the current tip passes **all** checks is probably the one that most people want. In fact I would argue that this could be the default "poster-boy-badge" for GitHub checks, as it also corresponds to the checkmark next to the latest commit that is prominently displayed on the repo front page.

**Comment by calebcartwright** [comment]
> Personally I think badges that show number of checks or show passed/failed checks is pretty niche.
>
> In contrast, having a badge that simply shows if the current tip passes all checks is probably the one that most people want. In fact I would argue that this could be the default "poster-boy-badge" for GitHub checks, as it also corresponds to the checkmark next to the latest commit that is prominently displayed on the repo front page.

Thanks for sharing your perspective. Would be good to hear from others.

Personally, I don't think it's niche to offer an additional level of detail. I think having a badge display the same data point  that GitHub already displays itself not too many pixels above is more niche :smile: To each their own though

**Comment by daxpedda** [comment]
> I think having a badge display the same data point that GitHub already displays itself not too many pixels above is more niche 😄

Really? Basically every repo I know that has badges in their README uses the regular CI badge, which only shows a specific workflow, and not all. I'm pretty sure the majority of repos didn't intend that and would prefer the badge to show if the whole CI currently succeeds or not, it's just whats currently available.

The green checkmark next to the latest commit is not very visible, having the badge in the README is way more eye-catching. In addition the README is often re-distributed to other platforms where the green checkmark doesn't exist, but the regular CI badge can't reflect the same status.

**Comment by calebcartwright** [comment]
> Really? Basically every repo I know that has badges in their README uses the regular CI badge, which only shows a specific workflow, and not all. I'm pretty sure the majority of repos didn't intend that and would prefer the badge to show if the whole CI currently succeeds or not, but they have to make due with whats currently available.
>
> EDIT: I'm obviously making a lot of assumptions here I can't really verify ...

Yeah, I don't want to get into the business of claiming to know what's in the head of another human, but I'd dispute any implications/characterizations that most people mistake our GitHub Workflow Status badge and GitHub's own native Workflow Status badge as being more holistic than a specific workflow (they have to provide a specific workflow name/workflow file name to get the badge, after all).

Inevitably I'm sure some make the mistake, but I'm going to give the broader global engineering community the benefit of the doubt that they understand that distinction.

This, however, is orthogonal to the topic at hand of what users would _like_, and I'd rather we steer discussion back in that direction

**Comment by calebcartwright** [comment]
I'm also trying to reload some mental context around GitHub Actions/Checks/Check Suites.

Is it accurate that, e.g. if a repo has 3 GHA workflows, each with 3 jobs as well as a 3rd party integration (let's say something like Codecov or Coveralls that provides coverage info) that there would be 10 total checks showing in the response for this API?

If so then I can understand the desire for an aggregate status, including the succinct one (again these are not mutually exclusive, we have lots of badges that support both brief and more verbose message details). It's just that the requests we'd received previously were for the reverse: more granular/detailed statuses for a specific check or subsets

**Comment by daxpedda** [comment]
> Inevitably I'm sure some make the mistake, but I'm going to give the broader global engineering community the benefit of the doubt that they understand that distinction.

Just to clarify, I meant that people make due with whats available, not that they misinterpret what it does. Apologies for all the edits!

> Is it accurate that, e.g. if a repo has 3 GHA workflows, each with 3 jobs as well as a 3rd party integration (let's say something like Codecov or Coveralls that provides coverage info) that there would be 10 total checks showing in the response for this API?

That is indeed my current understanding of this API.

**Comment by calebcartwright** [comment]
Thanks for confirming. I have a better understanding of the use case and benefits now, and am generally a :+1: to moving forward with this.

One thing we'll definitely need to do is extend the default github badge documentation to note the limitation around external/fork refs (to attempt to head off some of the inevitable "why isn't this working for my fork/PR" questions).

I also want to think about the route a bit more to make sure that we don't box ourselves into a corner (i.e. so that in the future we can naturally support a variant with the more detailed summary style messages, a status for a specific, individual check, etc.)

**Comment by mbtools** [comment]
### Clarification / Tests

I did some tests and the note in the GH docs is not correct or at least misleading. The API does return the check-runs for PRs that originate from a fork. 

Here's an example: 
- [PR from a fork](https://github.com/Marc-Bernard-Tools/ABAP-Tools-for-abapGit/pull/1) with 2 commits
- [API response for first commit](https://api.github.com/repos/Marc-Bernard-Tools/ABAP-Tools-for-abapGit/commits/44ab24f76197a09cfad46aa68fc065f216be749c/check-runs) - one check with failure, three with success
- [API response for second commit](https://api.github.com/repos/Marc-Bernard-Tools/ABAP-Tools-for-abapGit/commits/fe3674a19fea2c940dd8925e34bc6fbc5e6b896d/check-runs) - all checks successfully completed

If you just push something to a fork without a PR, there's no commit in the original repo and therefore no corresponding check-runs in that repo. 

Overall, it should very much meet user expectations.

### Aggregate / Detailed State

Adding an option to show details like "1 failed, 3 passed" makes sense. I just wanted to get something working first 😃 but please do suggest improvements to the route.

**Comment by calebcartwright** [comment]
Awesome, thank you for confirming! I'll try to propose some route changes later today

**Comment by mbtools** [comment]
Can we aim to merge this as a Christmas present, please? 🎄 🎁 😀

**Comment by PyvesB** [comment]
@calebcartwright any chance you could circle back to this one, or shall @chris48s or myself take a look? :)

**Comment by hikerspath** [comment]
Anyone home Github?  This has been an open request since 2019, someone finally gives you the code and we're coming up on 2 years with no merge?

**Comment by hikerspath** [comment]
The author of this request has complied with all of the rule changes and other asks around getting status badges for check runs, yet here we are still from 2019 (#4364) waiting.

**Value**: There are many third-party integrations that do not provide status checks.  This means that while it is possible to integrate with them, it is not possible to gain any status reporting from them.  Status on 'check_run' would provide exactly this.

It would be epic if we could honor the work done by @mbtools and get this PR merged before this turns 5.

**Comment by mbtools** [comment]
Ladies and Gentlemen, Dear Maintainers,

it's been 2 years since I opened this PR. Do I qualify as the most patient contributor on GitHub yet? 

I will give it one more shot. Please review and let me know what should be changed. If I do not get any reaction soon, I will close this PR.

Best,
Marc

**Comment by bourgeoisor** [comment]
I've been watching this PR for over a year now, hoping it gets merged. A little bit of communication from the maintainers would be great to get this going. If there's changes that needs to happen before this is merged in, then let's spell out those changes so they can be fixed.

**Comment by PyvesB** [comment]
As a reminder, this is an open-source project, a small team of maintainers dedicate their spare time for free, some of us having diligently done so since 2017. If folks want this so badly, why don't they participate in a constructive way, for example by doing a first review pass, testing the new badges, and reporting on their findings here? Implying that there's been a five year wait, for an issue that saw outside activity for the first time in 2022, feels all the more out of place.

@mbtools we do value your contribution and the time you put into this, and I'm sorry we've kept you on the hook for so long. I am a little stretched at the moment, but I do want to pick this up at some point. If you could extend your patience a little further, that would be great! 😃

**Comment by mbtools** [comment]
No worries, I will leave it open. Thanks for putting it back on the radar 👍

**Comment by PyvesB** [comment]
The deployment to a review app to test this out has failed, hopefully #10134 should fix it.

**Comment by mbtools** [comment]
I added an optional name filter so you can get the status of one particular check run

**Comment by mbtools** [comment]
image

**Review by PyvesB — COMMENTED** [review]
Thanks for the good work, one small change and I'd say we can merge this! :)

File: services/github/github-check-runs.service.js
Comment (PyvesB) [reply services/github/github-check-runs.service.js:106]: The following line ignores it dynamically, but probably best to not include it in the badge definition:
https://github.com/badges/shields/blob/d68217295114c653d0a803463f368df3f146acda/core/base-service/coalesce-badge.js#L153

Comment (mbtools) [reply services/github/github-check-runs.service.js:106]: Hmm. It won't use the logo unless "style" is set to "social". There are also several other GH badges with the same:

image

Still remove it?

Comment (PyvesB) [reply services/github/github-check-runs.service.js:106]: I'd say so, as mentioned, I'm proposing a cleanup pass in #10195.

**Review by PyvesB — APPROVED** [review]

**Comment by PyvesB** [comment]
Thanks for your patience on this one @mbtools! Could you please prepare the follow up PR we've been discussing in https://github.com/badges/shields/issues/4364#issuecomment-2106274450? 🙏🏻

**Comment by mbtools** [comment]
Woohoo! I will have some 🍰 and pop a bottle of 🍾 🥳 🎈 🎉 

PS: I will get back to the comment later today

**Pull request merged**
